web/pdf_thumbnail_view.js

```diff
@@ -156,296 +148,282 @@ var PDFThumbnailView = (function PDFThumbnailViewClosure() {
-      var img = pageView.canvas;
-      if (!img) {
+
+      if (((typeof PDFJSDev === 'undefined' ||
+            !PDFJSDev.test('PDFJS_NEXT')) && error === 'cancelled') ||
+          error instanceof RenderingCancelledException) {
+        renderCapability.resolve(undefined);
         return;
       }
-      if (!this.pdfPage) {
-        this.setPdfPage(pageView.pdfPage);
-      }
 
       this.renderingState = RenderingStates.FINISHED;
+      this._convertCanvasToImage();
 
-      var ctx = this._getPageDrawContext(true);
-      var canvas = ctx.canvas;
+      if (!error) {
+        renderCapability.resolve(undefined);
+      } else {
+        renderCapability.reject(error);
+      }
+    };
 
-      if (img.width <= 2 * canvas.width) {
-        ctx.drawImage(img, 0, 0, img.width, img.height,
-                      0, 0, canvas.width, canvas.height);
-        this._convertCanvasToImage();
+    let ctx = this._getPageDrawContext();
+    let drawViewport = this.viewport.clone({ scale: this.scale, });
+    let renderContinueCallback = (cont) => {
+      if (!this.renderingQueue.isHighestPriority(this)) {
+        this.renderingState = RenderingStates.PAUSED;
+        this.resume = () => {
+          this.renderingState = RenderingStates.RUNNING;
+          cont();
+        };
         return;
       }
-      // drawImage does an awful job of rescaling the image, doing it gradually.
-      var MAX_NUM_SCALING_STEPS = 3;
-      var reducedWidth = canvas.width << MAX_NUM_SCALING_STEPS;
-      var reducedHeight = canvas.height << MAX_NUM_SCALING_STEPS;
-      var reducedImage = TempImageFactory.getCanvas(reducedWidth,
-                                                    reducedHeight);
-      var reducedImageCtx = reducedImage.getContext('2d');
-
-      while (reducedWidth > img.width || reducedHeight > img.height) {
-        reducedWidth >>= 1;
-        reducedHeight >>= 1;
-      }
-      reducedImageCtx.drawImage(img, 0, 0, img.width, img.height,
-                                0, 0, reducedWidth, reducedHeight);
-      while (reducedWidth > 2 * canvas.width) {
-        reducedImageCtx.drawImage(reducedImage,
-                                  0, 0, reducedWidth, reducedHeight,
-                                  0, 0, reducedWidth >> 1, reducedHeight >> 1);
-        reducedWidth >>= 1;
-        reducedHeight >>= 1;
-      }
-      ctx.drawImage(reducedImage, 0, 0, reducedWidth, reducedHeight,
+      cont();
+    };
+
+    let renderContext = {
+      canvasContext: ctx,
+      viewport: drawViewport,
+    };
+    let renderTask = this.renderTask = this.pdfPage.render(renderContext);
+    renderTask.onContinue = renderContinueCallback;
+
+    renderTask.promise.then(function() {
+      finishRenderTask(null);
+    }, function(error) {
+      finishRenderTask(error);
+    });
+    return renderCapability.promise;
+  }
+
+  setImage(pageView) {
+    let img = pageView.canvas;
+    if (!img || this.renderingState !== RenderingStates.INITIAL) {
```

Comment:
Given that these two conditions are quite different, I'm wondering if having two different `if`s (as in the current code) actually aid readability somewhat here!?